### PR TITLE
CI Job: Compile and execute

### PR DESCRIPTION
**Pull request by casperboone**
This PR adds compilation of tests and (/with) library.

It was a bit of a struggle to get everything to work together, mostly because of super nice class path issues. But everything should work now, yay! 🎉 

The code in this PR kind of assumes that everything will work without any issues. Nice handling for errors during for instance compilation or even downloading the zip, will be added later. Also, tests for that kind of 'bad weather behavior' will be added then.

To improve the real lifeness of the tests, I created a dummy project repository at https://github.com/cafejojo/dummy-simple-maven-project.

### CODE REVIEW

File: modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt [hunk 1]
```diff
@@ -3,22 +3,37 @@ package org.cafejojo.schaapi.validationpipeline.cijob
...
-    fun run() {
-        download()
-        extractZip()
+    fun run(): TestResults {
+        if (!testsDirectory.exists()) throw CIJobException("This project has not been initialized yet. $testsDirectory")
+
+        next("Start downloading...") { download() }
+        next("Extract zip...") { extractZip() }
+        val library = next("Start compile library...") { compileLibrary() }
+        val tests = next( "Start compile tests...") { compileTests(library) }
```
Comment (gandreadis) [inline modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt]: `compiling` perhaps fits better here?

File: modules/validation-pipeline/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/TestRunner.kt [hunk 2]
```diff
@@ -11,9 +11,10 @@ interface TestRunner {
      *
      * @param rootDir the root directory of the project that contains the test files
      * @param testFiles the files that contain the tests to be run
+     * @param classPathDirectories the directories that need to be put on the class path during a test execution
```
Comment (gandreadis) [inline modules/validation-pipeline/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/TestRunner.kt]: `classPath` should be `classpath`, to be consistent with the rest of the codebase

**Comment by gandreadis** [comment]
Also, the tests seem to be failing - did you perhaps forget to check in the JUnit JAR?

**Comment by casperboone** [comment]
@gandreadis I did, but that does not seem to be the reason of the failure. It apparently cannot find the resource directory 🤔.

Anyway, I fixed the two issues you made a comment about :)

File: modules/application/src/main/kotlin/org/cafejojo/schaapi/CommandLineInterface.kt [hunk 3]
```diff
@@ -124,12 +124,6 @@ private fun parseArgs(options: Options, args: Array<String>): CommandLine? {
     }
 }
 
-private fun installMavenIfNeeded(mavenDir: File, cmd: CommandLine) {
-    if (!mavenDir.resolve("bin/mvn").exists() || cmd.hasOption("repair_maven")) {
-        MavenInstaller().installMaven(mavenDir)
-    }
-}
```
Comment (FWDekker) [inline modules/application/src/main/kotlin/org/cafejojo/schaapi/CommandLineInterface.kt]: Why though?

File: modules/mining-pipeline/java-maven-project-compiler/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/projectcompiler/javamaven/MavenInstaller.kt [hunk 4]
```diff
@@ -12,7 +12,9 @@ class MavenInstaller {
      *
      * @param path the directory to install Maven in
      */
-    fun installMaven(path: File) {
+    fun installMaven(path: File, overwrite: Boolean = false) {
+        if (path.resolve("bin/mvn").exists() && !overwrite) return
```
Comment (FWDekker) [inline modules/mining-pipeline/java-maven-project-compiler/src/main/kotlin/org/cafejojo/schaapi/miningpipeline/projectcompiler/javamaven/MavenInstaller.kt]: This feels somewhat out of scope for this PR. Why didn't you comment these suggestions on #250?

File: modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt [hunk 5]
```diff
@@ -3,22 +3,40 @@ package org.cafejojo.schaapi.validationpipeline.cijob
...
-        download()
-        extractZip()
+    fun run(): TestResults {
+        if (!testsDirectory.exists()) throw CIJobException("This project has not been initialized yet. $testsDirectory")
+
+        MavenInstaller().installMaven(JavaMavenProject.DEFAULT_MAVEN_HOME)
+
+        next("Start downloading...") { download() }
+        next("Extract zip...") { extractZip() }
+        val library = next("Start compiling library...") { compileLibrary() }
```
Comment (FWDekker) [inline modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt]: I think you should either remove "Start" or remove "...", because _starting_ to compile the library does not take much time

File: modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt [hunk 6]
```diff
@@ -27,7 +45,36 @@ class CIJob(private val identifier: String, private val projectDirectory: File,
...
+
+        val classPath = mutableListOf<String>().run {
+            addAll(testsDirectory.listFiles().filter { it.extension == "class" }.map { it.parent })
+            add(library.classDir.absolutePath)
+            add(CIJob::class.java.getResource("/junit-4.12.jar").file)
+
+            joinToString(File.pathSeparator)
+        }
+
+        return testsDirectory.listFiles().filter { it.extension == "java" }
```
Comment (FWDekker) [inline modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt]: Could you put `filter` on the next line?

File: modules/validation-pipeline/junit-test-runner/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/testrunner/junit/TestRunner.kt [hunk 7]
```diff
@@ -30,7 +33,11 @@ class TestRunner : TestRunner {
                     val className = testFile.toRelativeString(rootDir)
                         .replace(Regex("[/\\\\]"), ".")
                         .removeSuffix(".class")
-                    className to JUnitCore.runClasses(this.javaClass.classLoader.loadClass(className))
+                    val classLoader = URLClassLoader(
+                        classpathDirectories.map { it.toURI().toURL() }.toTypedArray()
```
Comment (FWDekker) [inline modules/validation-pipeline/junit-test-runner/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/testrunner/junit/TestRunner.kt]: `it.toURI().toURL()`

i cri everitim
#backwards-compatibility

File: modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt [hunk 8]
```diff
@@ -3,22 +3,40 @@ package org.cafejojo.schaapi.validationpipeline.cijob
...
      */
-    fun run() {
-        download()
-        extractZip()
+    fun run(): TestResults {
+        if (!testsDirectory.exists()) throw CIJobException("This project has not been initialized yet. $testsDirectory")
+
+        MavenInstaller().installMaven(JavaMavenProject.DEFAULT_MAVEN_HOME)
+
+        next("Downloading...") { download() }
```
Comment (JSAbrahams) [inline modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt]: Nice

File: modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt [hunk 9]
```diff
@@ -27,7 +45,36 @@ class CIJob(private val identifier: String, private val projectDirectory: File,
...
+            addAll(testsDirectory.listFiles().filter { it.extension == "class" }.map { it.parent })
+            add(library.classDir.absolutePath)
+            add(CIJob::class.java.getResource("/junit-4.12.jar").file)
+
+            joinToString(File.pathSeparator)
+        }
+
+        return testsDirectory.listFiles().filter { it.extension == "java" }
+            .also { files ->
+                if (files.any { compiler.run(null, null, errorOutput, it.absolutePath, "-cp", classPath) != 0 }) {
```
Comment (JSAbrahams) [inline modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt]: I'd make this its own private method for readability

File: modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt (hunk 6 above)
Comment (JSAbrahams) [reply modules/validation-pipeline/ci-job/src/main/kotlin/org/cafejojo/schaapi/validationpipeline/cijob/CIJob.kt]: It is more consistent with the `also` and `map` below
